Numpy

In [2]:
import numpy as np

In [4]:
arr = np.array([16,12,5,4])
print(type(arr))
print(arr)

<class 'numpy.ndarray'>
[16 12  5  4]


In [5]:
arr = np.array((16,12,5,4))
print(type(arr))
print(arr)

<class 'numpy.ndarray'>
[16 12  5  4]


In [7]:
import numpy as np

arr_1 = np.array([20, 24, 16, 5])
arr_2 = np.array([5, 8, 2, 4])

print("Addition:", arr_1 + arr_2)
print("Subtraction:", arr_1 - arr_2)
print("Multiplication:", arr_1 * arr_2)
print("Division:", arr_1 / arr_2)
print("Modulus:", arr_1 % arr_2)

Addition: [25 32 18  9]
Subtraction: [15 16 14  1]
Multiplication: [100 192  32  20]
Division: [4.   3.   8.   1.25]
Modulus: [0 0 0 1]


In [8]:
arr = np.array([16, 4, 25, 64])

print(np.sum(arr))    
print(np.sqrt(arr))    
print(np.min(arr))     
print(np.max(arr))    
print(np.mean(arr))    

109
[4. 2. 5. 8.]
4
64
27.25


In [11]:
arr = np.array([1, 2, 3, 4, 5], dtype=np.int32) #int32 = integer stored using 4 bytes (32 bits)
print(arr)
print(arr.dtype)

[1 2 3 4 5]
int32


In [12]:
arr = np.array([1, 2, 3, 4, 5], dtype=np.int64) #int64 = integer stored using 4 bytes (32 bits)
print(arr)
print(arr.dtype)

[1 2 3 4 5]
int64


In [13]:
arr = np.array([1, 2, 3, 4, 5], dtype=np.int8) #int32 = integer stored using 4 bytes (32 bits)
print(arr)
print(arr.dtype)

[1 2 3 4 5]
int8


 

`dtype` in NumPy defines how data is stored in memory.

We use different types like `int8`, `int16`, `int32`, and `int64` to:

* Control memory usage (smaller types use less memory)
* Improve performance (smaller data types can be faster)
* Control number range (avoid overflow or invalid values)
* Optimize real-world applications (images, audio, data science)

In simple terms:
`dtype` decides how NumPy stores and handles numbers efficiently in memory.


In [14]:
# float32 = 32-bit floating point (4 bytes)
arr = np.array([1, 2, 3, 4, 5], dtype=np.float32)
print(arr)
print(arr.dtype)

# float64 = 64-bit floating point (8 bytes)
arr = np.array([1, 2, 3, 4, 5], dtype=np.float64)
print(arr)
print(arr.dtype)

[1. 2. 3. 4. 5.]
float32
[1. 2. 3. 4. 5.]
float64


float32 → uses less memory (4 bytes) and stores decimal numbers with lower precision

float64 → uses more memory (8 bytes) and stores decimal numbers with higher precision (higher accuracy)

Both represent decimal numbers, but they differ in how accurately the values are stored in memory.

**2D Array

A 2D array is a collection of data arranged in rows and columns (like a table).

Example:

import numpy as np

arr = np.array([
    [1, 2, 3],
    [4, 5, 6]
])
print(arr)

  Used to represent tables, Excel sheets, and simple matrices.

**3D Array

A 3D array is a collection of multiple 2D arrays (like stacked tables).

Example:

import numpy as np

arr = np.array([
    [[1, 2], [3, 4]],
    [[5, 6], [7, 8]]
])
print(arr)

Used in images (RGB), videos, and advanced data structures.

Difference
2D → rows and columns (table)
3D → multiple 2D tables stacked together

2D and 3D Arrays

In [17]:
sample_2d_array = np.array([
    [1, 2, 3],
    [4, 5, 6]
])  # 2 rows, 3 columns

print(sample_2d_array)

[[1 2 3]
 [4 5 6]]


In [20]:
sample_2d_array = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

print("Array:")
print(arr)

print("Shape:", sample_2d_array.shape)
print("Size:", sample_2d_array.size)
print("Dimensions (ndim):", sample_2d_array.ndim)

Array:
[[1 2 3]
 [4 5 6]]
Shape: (2, 3)
Size: 6
Dimensions (ndim): 2


In [21]:
#(2,3,4)
arr = np.array([
    [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12]
    ],
    [
        [13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]
    ]
])

print(arr)
print("Shape:", arr.shape)



[[[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]

 [[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]]
Shape: (2, 3, 4)


What is a Tensor?

A tensor is a general term for multi-dimensional data:

0D → scalar (single number)
1D → vector (list)
2D → matrix (table)
3D → tensor (stack of matrices)
4D+ → higher-dimensional tensors

In [26]:
#(2,3,4)
arr = np.array([
    [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12]
    ],
    [
        [13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]
    ]
])

print(arr)
print("Shape:", arr.shape)
print(arr[1][0][2])# not preper , slow , don't use
print(arr[1,0,2]) # use this , It is correct way
print(arr[-1,-3,-2])




[[[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]

 [[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]]
Shape: (2, 3, 4)
15
15
15


Boolean Comparison

In [27]:
arr = np.array([10,30,20,15,5])

condition = arr > 15
print(type(condition))
print(condition)

<class 'numpy.ndarray'>
[False  True  True False False]


In [29]:
arr = np.array([10,30,20,15,5])

condition = arr > 15
print(type(condition))
print(condition)

filtered_arr = arr[condition]
print(filtered_arr)
print(type(filtered_arr))

<class 'numpy.ndarray'>
[False  True  True False False]
[30 20]
<class 'numpy.ndarray'>


In [30]:
filtered_arr = arr[arr > 15]
print(filtered_arr)

[30 20]


In [33]:
arr = np.array([10,30,20,15,5])

filtered_arr = arr[(arr < 25) & (arr > 10)]
print(filtered_arr)

[20 15]


Numpy Array Initialization

In [34]:
#np.zeros(shape)
sample_array = np.zeros(5)
print(sample_array)

[0. 0. 0. 0. 0.]


In [35]:
#np.zeros(shape)
sample_array = np.zeros((2,3))
print(sample_array)

[[0. 0. 0.]
 [0. 0. 0.]]


In [37]:
#np.zeros(shape)
sample_array = np.ones((2,3))
print(sample_array)

[[1. 1. 1.]
 [1. 1. 1.]]


In [39]:
#np.zeros(shape)
# np.full(shape, fill_value)
sample_array = np.full((2,3),7)
print(sample_array)

[[7 7 7]
 [7 7 7]]


In [42]:
#np.zeros(shape)
# np.full(shape, fill_value)
sample_array = np.full((2,3),[1,8,12])
print(sample_array)

[[ 1  8 12]
 [ 1  8 12]]


In [ ]:
sample_array = np.empty((2, 3))
print(sample_array)  # output - grabage value fill 
# very fast, future help for value assign

[[5.e-324 4.e-323 6.e-323]
 [5.e-324 4.e-323 6.e-323]]


Pandas Library

Pandas is a Python library for working with tabular data (rows and columns).

It is used to store, process, and analyze data.

It can handle data from Excel and CSV files.

It is an open-source library.

In [45]:
import pandas as pd

Pandas Series vs DataFrame – Summary Note

Pandas Series

1-dimensional data structure

Like a single column

Has index for each value

Used for simple data storage

Pandas DataFrame

2-dimensional data structure

Like a table (rows and columns)

Combination of multiple Series

Used for structured data analysis

Difference

Series → 1D (single column)
DataFrame → 2D (table format)

In [ ]:
tempertures = pd.Series([32,28,35,30])

print(tempertures)
print(type(tempertures))

# 0,1,2,3 → labels
# 32,28,35,30 → values (0,1,2,3 - index stores)
# dtype: int64 → values are stored as 64-bit integers
# <class 'pandas.core.series.Series'> → the object is a Pandas Series

0    32
1    28
2    35
3    30
dtype: int64
<class 'pandas.core.series.Series'>


Label Access

In [47]:
tempertures = pd.Series([32,28,35,30], index=['Monday','Tuesday','Wendsday','Thusday'])

print(tempertures)
print(type(tempertures))

Monday      32
Tuesday     28
Wendsday    35
Thusday     30
dtype: int64
<class 'pandas.core.series.Series'>


 

loc	   -            Label-based indexing

iloc	-           Integer index-based indexing

In [48]:
 
tempertures = pd.Series([32,28,35,30], index=['Monday','Tuesday','Wendsday','Thusday'])

print(tempertures)
print(type(tempertures))
print(tempertures.loc['Wendsday'])
print(tempertures.iloc[2])

Monday      32
Tuesday     28
Wendsday    35
Thusday     30
dtype: int64
<class 'pandas.core.series.Series'>
35
35


In [3]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 35],
    'Gender': ['Female', 'Male', 'Male']
}

df = pd.DataFrame(data)

print(df)
print(type(df))

      Name  Age  Gender
0    Alice   25  Female
1      Bob   30    Male
2  Charlie   35    Male
<class 'pandas.core.frame.DataFrame'>


In [5]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 35],
    'Gender': ['Female', 'Male', 'Male']
}

df = pd.DataFrame(data)

print(df)
print(type(df))

      Name  Age  Gender
0    Alice   25  Female
1      Bob   30    Male
2  Charlie   35    Male
<class 'pandas.core.frame.DataFrame'>


In [4]:
data = {
    'Name': ['Alice', 'Bob'],
    'Age': [25, 30],
    'Gender': ['Female', 'Male']
}

df = pd.DataFrame(data, index=['student_1', 'student_2'])

print(df)
print(type(df))

            Name  Age  Gender
student_1  Alice   25  Female
student_2    Bob   30    Male
<class 'pandas.core.frame.DataFrame'>


In [17]:
data = {
    'Name': ['Alice', 'Bob','Charlie'],
    'Age': [25, 30, 35],
    'Gender': ['F', 'M', 'M']
}

df = pd.DataFrame(data, index=['student_1', 'student_2', 'student_3'])

# student_2 row
print(df.loc['student_2'])
print(type(df.loc['student_2']))
print("-------------")

# Age column
print(df['Age'])
print(type(df['Age']))
print("-------------")

# value 30
print(df.loc['student_2', 'Age'])
print(type(df.loc['student_2', 'Age']))

Name      Bob
Age        30
Gender      M
Name: student_2, dtype: object
<class 'pandas.core.series.Series'>
-------------
student_1    25
student_2    30
student_3    35
Name: Age, dtype: int64
<class 'pandas.core.series.Series'>
-------------
30
<class 'numpy.int64'>


A DataFrame row is converted into a Series because it becomes a single 1D object with labels (column names).

In [18]:

# shape (rows, columns)
print("Shape:", df.shape)

# size (total elements)
print("Size:", df.size)

Shape: (3, 3)
Size: 9


In [21]:
len(df)
# numbers of rows

3

In [22]:
print(df.dtypes)

Name      object
Age        int64
Gender    object
dtype: object


In [24]:
df["Name"] = df["Name"].astype("string")

print(df.dtypes)

Name      string[python]
Age                int64
Gender            object
dtype: object


In [25]:
column_names = df.columns
print(column_names)
print(type(column_names))

Index(['Name', 'Age', 'Gender'], dtype='object')
<class 'pandas.core.indexes.base.Index'>


In [26]:
print(list(column_names))

['Name', 'Age', 'Gender']


In [ ]:
print(df.values)
print(type(df.values))

[['Alice' 25 'F']
 ['Bob' 30 'M']
 ['Charlie' 35 'M']]
<class 'numpy.ndarray'>


In [29]:
df

,Name,Age,Gender
student_1,Alice,25,F
student_2,Bob,30,M
student_3,Charlie,35,M


In [33]:
data = {
    'Age': [25, 30, 35, 40],
    'Salary': [50000, 60000, 70000, 80000]
}

df = pd.DataFrame(data)
print(df)
df

   Age  Salary
0   25   50000
1   30   60000
2   35   70000
3   40   80000


,Age,Salary
0,25,50000
1,30,60000
2,35,70000
3,40,80000


sum() → total values

mean() → average values

describe() → full statistics report

Works mainly on numeric columns in DataFrame

In [37]:
#.sum() - 
#.mean()
#.describe()
data = {
    'Age': [25, 30, 35, 40],
    'Salary': [50000, 60000, 70000, 80000]
}

df = pd.DataFrame(data)
print(df.sum())
print("--------")
print(df.mean())
print("-----------")
print(df.describe())



Age          130
Salary    260000
dtype: int64
--------
Age          32.5
Salary    65000.0
dtype: float64
-----------
             Age        Salary
count   4.000000      4.000000
mean   32.500000  65000.000000
std     6.454972  12909.944487
min    25.000000  50000.000000
25%    28.750000  57500.000000
50%    32.500000  65000.000000
75%    36.250000  72500.000000
max    40.000000  80000.000000


In [36]:
print(df.describe())

             Age        Salary
count   4.000000      4.000000
mean   32.500000  65000.000000
std     6.454972  12909.944487
min    25.000000  50000.000000
25%    28.750000  57500.000000
50%    32.500000  65000.000000
75%    36.250000  72500.000000
max    40.000000  80000.000000


Standard deviation = measure of data spread around the mean

Standard Deviation and Quartiles (Pandas Describe Output)
Standard Deviation (std)

Standard deviation measures how much the data values are spread around the mean (average).

From your output:

Age std = 6.45
Salary std = 12909.94

Meaning:

Age values are less spread around the mean
Salary values are more spread around the mean
Quartiles

Quartiles divide the data into four equal parts.

25% (Q1 - First Quartile)
Age = 28.75
Salary = 57500

Meaning:
25% of the data values are below this point.

50% (Q2 - Median)
Age = 32.5
Salary = 65000

Meaning:
50% of the data values are below this point (middle value).

75% (Q3 - Third Quartile)
Age = 36.25
Salary = 72500

Meaning:
75% of the data values are below this point.

Summary Table
Term	Meaning
std	Measures data spread around the mean
25% (Q1)	First quarter of data
50% (Q2)	Median (middle value)fi
75% (Q3)	Third quarter of data
Final Idea

Standard deviation explains how spread the data is, and quartiles divide the dataset into four equal parts based on position.

In [43]:
filtered_arr = df[df['Age'] > 25]
filtered_arr

,Age,Salary
1,30,60000
2,35,70000
3,40,80000


In [42]:
filtered_arr = df[(df['Age'] > 30) & (df['Salary'] < 75000)]
print(filtered_arr)

   Age  Salary
2   35   70000


In [44]:
sorted_df = df.sort_values(by='Age')
print(sorted_df)

   Age  Salary
0   25   50000
1   30   60000
2   35   70000
3   40   80000


In [45]:
sorted_df = df.sort_values(by='Salary')
print(sorted_df)

   Age  Salary
0   25   50000
1   30   60000
2   35   70000
3   40   80000


In [48]:
# add the new column - distance , km - 8,4,20,17
# add a neew row - 38, 78000 , 15

df['Distance_km'] = [8, 4, 20, 17]
df

 

,Age,Salary,Distance_km
0,25,50000,8
1,30,60000,4
2,35,70000,20
3,40,80000,17


In [49]:
df.loc[4] = [38, 78000, 15]
df

,Age,Salary,Distance_km
0,25,50000,8
1,30,60000,4
2,35,70000,20
3,40,80000,17
4,38,78000,15


Data Manipulation
Droppng a coloumn/row

df.drop(labels=,axis=1,inplace=False)

drop() → removes data

labels → what you want to delete (row or column name)
axis=1 → delete a column
axis=0 → delete a row
inplace=True → modifies the original DataFrame directly  

inplace=False (default) → safe, returns new data
inplace=True → changes original data

In [51]:
new_df = df.drop('Distance_km', axis=1)
print(new_df)

   Age  Salary
0   25   50000
1   30   60000
2   35   70000
3   40   80000
4   38   78000


In [52]:
df = pd.read_csv('sample.csv')
print(df)
print(type(df))


    id           name  age           city  salary
0    1  Alice Johnson   28       New York   55000
1    2      Bob Smith   34    Los Angeles   62000
2    3  Charlie Brown   45        Chicago   78000
3    4   Diana Prince   29        Houston   59000
4    5     Ethan Hunt   38        Phoenix   71000
5    6    Fiona Apple   26   Philadelphia   52000
6    7  George Miller   52    San Antonio   85000
7    8     Hannah Lee   31      San Diego   64000
8    9     Ian Wright   40         Dallas   73000
9   10  Julia Roberts   35       San Jose   68000
10  11     Kevin Hart   33         Austin   66000
11  12   Laura Palmer   27   Jacksonville   54000
12  13  Michael Scott   48     Fort Worth   82000
13  14    Nina Simone   30       Columbus   61000
14  15    Oscar Wilde   42      Charlotte   75000
15  16    Paula Abdul   36  San Francisco   69000
16  17   Quincy Jones   55   Indianapolis   90000
17  18   Rachel Green   32        Seattle   67000
18  19   Steve Rogers   41         Denver   74000


In [ ]:

df = pd.read_json()
print(df)